File: Projetos/ACBrLib/Demos/C#/Shared/ACBrLib.Core/Ini/ACBrIniFile.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace ACBrLib.Core
{
    public sealed class ACBrIniFile : IEnumerable<ACBrIniSection>
    {
        #region Fields

        private List<ACBrIniSection> sections;

        #endregion Fields

        #region Constructors

        public ACBrIniFile() : this("", "")
        {
        }

        public ACBrIniFile(string iniFilePath, string iniFileName) : this(iniFilePath, iniFileName, Encoding.GetEncoding("ISO-8859-1"), 1024)
        {
        }

        public ACBrIniFile(string iniFilePath, string iniFileName, Encoding encoding, int bufferSize)
        {
            sections = new List<ACBrIniSection>();
            Encoding = encoding;
            BufferSize = bufferSize;
            IniFilePath = iniFilePath;
            IniFileName = iniFileName;
        }

        #endregion Constructors

        #region Properties

        public string IniFileName { get; set; }

        public string IniFilePath { get; set; }

        public Encoding Encoding { get; set; }

        public int BufferSize { get; set; }

        public int SectionCount => sections.Count;

        public ACBrIniSection this[int idx] => sections[idx];

        public ACBrIniSection this[string section]
        {
            get
            {
                var ret = sections.SingleOrDefault(x => x.Name == section);
                if (ret != null) return ret;

                ret = new ACBrIniSection(this, section);
                sections.Add(ret);

                return ret;
            }
        }

        #endregion Properties

        #region Methods

        /// <summary>
        /// Retorna true se a seção existir no ini.
        /// </summary>
        /// <param name="section"></param>
        /// <returns></returns>
        public bool Contains(string section)
        {
            return sections.Any(x => x.Name == section);
        }

        public bool Contains(ACBrIniSection section)
        {
            return Contains(section.Name);
        }

        public ACBrIniSection AddNew(string section)
        {
            if (Contains(section)) throw new ArgumentException("Sessão já existe no arquivo.");

            var ret = new ACBrIniSection(this, section);
            sections.Add(ret);
            return ret;
        }

        public void Add(ACBrIniSection section)
        {
            if (Contains(section.Name)) throw new ArgumentException("Sessão já existe no arquivo.");

            section.Parent = this;
            sections.Add(section);
        }

        public void Remove(string section)
        {
            var ret = sections.SingleOrDefault(x => x.Name == section);
            Remove(ret);
        }

        public void Remove(ACBrIniSection section)
        {
            if (section == null) return;
            Remove(section.Name);
            sections.Remove(section);
        }

        public TType Read<TType>(string section, string propertie, TType defaultValue = default)
        {
            if (string.IsNullOrEmpty(propertie) || string.IsNullOrWhiteSpace(propertie)) return defaultValue;
            if (string.IsNullOrEmpty(section) || string.IsNullOrWhiteSpace(section)) return defaultValue;

            var iniSection = this[section];
            return iniSection.GetValue(propertie, defaultValue);
        }

        public void Write(string section, string propertie, object value)
        {
            if (string.IsNullOrEmpty(propertie) || string.IsNullOrWhiteSpace(propertie)) return;
            if (string.IsNullOrEmpty(section) || string.IsNullOrWhiteSpace(section)) return;
            if (value == null) return;

            var iniSection = this[section];
            iniSection.SetValue(propertie, value);
        }

        public void Save()
        {
            if (string.IsNullOrEmpty(IniFilePath)) throw new ArgumentNullException();
            if (string.IsNullOrEmpty(IniFileName)) throw new ArgumentNullException();

            var file = Path.Combine(IniFilePath, IniFileName);

            using (var writer = new StreamWriter(file, false, Encoding, BufferSize))
                Save(writer);
        }

        public void Save(string file)
        {
            using (var writer = new StreamWriter(file, false, Encoding, BufferSize))
                Save(writer);
        }

        public void Save(Stream stream)
        {
            using (var writer = new StreamWriter(stream))
                Save(writer);
        }

        private void Save(TextWriter stream)
        {
            foreach (var section in sections)
                section.Save(stream);

            stream.Flush();
        }

        public static ACBrIniFile Parse(string iniData, Encoding encoding = null)
        {
            if (string.IsNullOrEmpty(iniData) || string.IsNullOrWhiteSpace(iniData)) throw new ArgumentNullException(nameof(iniData));

            encoding = encoding ?? Encoding.UTF8;
            var iniFile = new ACBrIniFile { Encoding = encoding };

            using (var reader = new StringReader(iniData))
            {
                string line;
                var section = string.Empty;
                while ((line = reader.ReadLine()) != null)
                {
                    line = line.Trim();

                    if (string.IsNullOrEmpty(line) || string.IsNullOrWhiteSpace(line)) continue;
                    if (line.StartsWith(";")) continue;

                    if (line.StartsWith("["))
                    {
                        section = line.Substring(1, line.Length - 2);
                        iniFile.sections.Add(new ACBrIniSection(iniFile, section));
                    }
                    else
                    {
                        if (string.IsNullOrEmpty(section) || string.IsNullOrWhiteSpace(section)) continue;

                        var iniSection = iniFile[section];
                        var idx = line.IndexOf('=', 0);
                        if (idx < 0) continue;

                        var key = line.Substring(0, idx).ToLower();
                        var value = idx >= line.Length - 1 ? "" : line.Substring(idx + 1);
                        iniSection.Add(key, value);
                    }
                }
            }

            return iniFile;
        }

        public static ACBrIniFile Load(string file, Encoding encoding = null)
        {
            if (!File.Exists(file)) throw new FileNotFoundException();

            using (var reader = new FileStream(file, FileMode.Open))
                return Load(reader, encoding);
        }

        public static ACBrIniFile Load(Stream stream, Encoding encoding = null)
        {
            if (stream == null) throw new ArgumentNullException(nameof(stream));

            encoding = encoding ?? Encoding.UTF8;
            var iniFile = new ACBrIniFile { Encoding = encoding };

            using (var reader = new StreamReader(stream, encoding))
            {
                string line;
                var section = string.Empty;
                while ((line = reader.ReadLine()) != null)
                {
                    line = line.Trim();

                    if (string.IsNullOrEmpty(line) || string.IsNullOrWhiteSpace(line)) continue;
                    if (line.StartsWith(";")) continue;

                    if (line.StartsWith("["))
                    {
                        section = line.Substring(1, line.Length - 2);
                        iniFile.sections.Add(new ACBrIniSection(iniFile, section));
                    }
                    else
                    {
                        if (string.IsNullOrEmpty(section) || string.IsNullOrWhiteSpace(section)) continue;

                        var iniSection = iniFile[section];
                        var idx = line.IndexOf('=', 0);
                        if (idx < 0) continue;

                        var key = line.Substring(0, idx);
                        var value = idx >= line.Length - 1 ? "" : line.Substring(idx + 1);
                        iniSection.Add(key, value);
                    }
                }
            }

            return iniFile;
        }

        public override string ToString()
        {
            var builder = new StringBuilder();

            using (var writer = new StringWriter(builder))
                Save(writer);

            return builder.ToString();
        }

        #endregion Methods

        #region IEnumerable

        /// <summary>
        /// Returns an enumerator that iterates through the collection.
        /// </summary>
        /// <returns>A <see cref="T:System.Collections.Generic.IEnumerator`1" /> that can be used to iterate through the collection.</returns>

        public IEnumerator<ACBrIniSection> GetEnumerator()
        {
            return sections.GetEnumerator();
        }

        /// <summary>
        /// Returns an enumerator that iterates through a collection.
        /// </summary>
        /// <returns>An <see cref="T:System.Collections.IEnumerator" /> object that can be used to iterate through the collection.</returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        #endregion IEnumerable
    }
}

File: Projetos/ACBrLib/Android/Comum/ACBrLibComum/src/main/java/br/com/acbr/lib/comum/ini/ACBrIniFile.java
package br.com.acbr.lib.comum.ini;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.Writer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class ACBrIniFile implements Iterable<ACBrIniSection> {

    private List<ACBrIniSection> sections;
    private String iniFilePath;
    private String iniFileName;
    private Charset encoding;
    private int bufferSize;

    public ACBrIniFile() {
        this("", "", StandardCharsets.ISO_8859_1, 1024);
    }

    public ACBrIniFile(String iniFilePath, String iniFileName) {
        this(iniFilePath, iniFileName, StandardCharsets.ISO_8859_1, 1024);
    }

    public ACBrIniFile(String iniFilePath, String iniFileName, Charset encoding, int bufferSize) {
        this.sections = new ArrayList<>();
        this.encoding = encoding;
        this.bufferSize = bufferSize;
        this.iniFilePath = iniFilePath;
        this.iniFileName = iniFileName;
    }

    public String getIniFileName() {
        return iniFileName;
    }

    public void setIniFileName(String iniFileName) {
        this.iniFileName = iniFileName;
    }

    public String getIniFilePath() {
        return iniFilePath;
    }

    public void setIniFilePath(String iniFilePath) {
        this.iniFilePath = iniFilePath;
    }

    public Charset getEncoding() {
        return encoding;
    }

    public void setEncoding(Charset encoding) {
        this.encoding = encoding;
    }

    public int getBufferSize() {
        return bufferSize;
    }

    public void setBufferSize(int bufferSize) {
        this.bufferSize = bufferSize;
    }

    public int getSectionCount() {
        return sections.size();
    }

    public ACBrIniSection getSection(int idx) {
        return sections.get(idx);
    }

    public ACBrIniSection getSection(String section) {
        for (ACBrIniSection sec : sections) {
            if (sec.getName().equals(section)) {
                return sec;
            }
        }
        ACBrIniSection newSection = new ACBrIniSection(this, section);
        sections.add(newSection);
        return newSection;
    }

    public boolean contains(String section){
        return sections.stream().anyMatch(s -> s.getName().equals(section));
    }

    public boolean contains(ACBrIniSection section) {
        return contains(section.getName());
    }

    public ACBrIniSection addNew(String section) {
        if (contains(section)) {
            throw new IllegalArgumentException("Sessão já existe no arquivo.");
        }
        ACBrIniSection newSection = new ACBrIniSection(this, section);
        sections.add(newSection);
        return newSection;
    }

    public void add(ACBrIniSection section) {
        if (contains(section.getName())) {
            throw new IllegalArgumentException("Sessão já existe no arquivo.");
        }
        section.setParent(this);
        sections.add(section);
    }

    public void remove(String section) {
        sections.removeIf(s -> s.getName().equals(section));
    }

    public void remove(ACBrIniSection section) {
        sections.remove(section);
    }

    public <T> T read(String section, String property, T defaultValue) {
        if (property == null || property.trim().isEmpty() || section == null || section.trim().isEmpty()) {
            return defaultValue;
        }
        ACBrIniSection iniSection = getSection(section);
        return iniSection.getValue(property, defaultValue);
    }

    public void write(String section, String property, Object value) {
        if (property == null || property.trim().isEmpty() || section == null || section.trim().isEmpty() || value == null) {
            return;
        }
        ACBrIniSection iniSection = getSection(section);
        iniSection.setValue(property, value);
    }

    public void save() throws IOException {
        if (iniFilePath == null || iniFileName == null) {
            throw new IllegalArgumentException("O caminho e o nome do arquivo devem ser especificados.");
        }
        File file = new File(iniFilePath, iniFileName);
        try (Writer writer = new OutputStreamWriter(new FileOutputStream(file), encoding)) {
            save(writer);
        }
    }

    public void save(String file) throws IOException {
        try (Writer writer = new OutputStreamWriter(new FileOutputStream(file), encoding)) {
            save(writer);
        }
    }

    public void save(OutputStream stream) throws IOException {
        try (Writer writer = new OutputStreamWriter(stream, encoding)) {
            save(writer);
        }
    }

    private void save(Writer writer) throws IOException {
        for (ACBrIniSection section : sections) {
            section.save(writer);
        }
        writer.flush();
    }

    public static ACBrIniFile parse(String iniData, Charset encoding) {
        if (iniData == null || iniData.trim().isEmpty()) {
            throw new IllegalArgumentException("Invalid ini data.");
        }
        ACBrIniFile iniFile = new ACBrIniFile();
        iniFile.encoding = encoding != null ? encoding : StandardCharsets.UTF_8;
        try (BufferedReader reader = new BufferedReader(new StringReader(iniData))) {
            String line;
            String currentSection = null;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith(";")) continue;
                if (line.startsWith("[")) {
                    currentSection = line.substring(1, line.length() - 1);
                    iniFile.sections.add(new ACBrIniSection(iniFile, currentSection));
                } else if (currentSection != null) {
                    String[] parts = line.split("=", 2);
                    String key = parts[0].trim();
                    String value = parts.length > 1 ? parts[1].trim() : "";
                    iniFile.getSection(currentSection).put(key, value);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return iniFile;
    }

    public static ACBrIniFile load(String file, Charset encoding) throws IOException {
        File f = new File(file);
        if (!f.exists()) throw new FileNotFoundException();
        try (InputStream stream = new FileInputStream(f)) {
            return load(stream, encoding);
        }
    }

    public static ACBrIniFile load(InputStream stream, Charset encoding) throws IOException {
        ACBrIniFile iniFile = new ACBrIniFile();
        iniFile.encoding = encoding != null ? encoding : StandardCharsets.UTF_8;
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, iniFile.encoding))) {
            String line;
            String currentSection = null;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith(";")) continue;
                if (line.startsWith("[")) {
                    currentSection = line.substring(1, line.length() - 1);
                    iniFile.sections.add(new ACBrIniSection(iniFile, currentSection));
                } else if (currentSection != null) {
                    String[] parts = line.split("=", 2);
                    String key = parts[0].trim();
                    String value = parts.length > 1 ? parts[1].trim() : "";
                    iniFile.getSection(currentSection).put(key, value);
                }
            }
        }
        return iniFile;
    }

    public String toString() {
        StringWriter writer = new StringWriter();
        try {
            save(writer);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return writer.toString();
    }

    @Override
    public Iterator<ACBrIniSection> iterator(){
        return sections.iterator();
    }

}
